Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,1481 @@
+
+theorem single_zero_right (a : G) : single a (0 : k) = 0 := by
+  ext; simp only [Finsupp.single_zero, Finsupp.coe_zero, Pi.zero_apply, toFinsupp_zero]
+
+@[simp]
+theorem single_add (a : G) (b₁ b₂ : k) : single a (b₁ + b₂) = single a b₁ + single a b₂ := by
+  ext; simp only [Finsupp.single_add, Finsupp.coe_add, Pi.add_apply, toFinsupp_add]
+
+theorem single_zero (a : G) : (single a 0 : SkewMonoidAlgebra k G) = 0 := by
+  rw [ofFinsupp_eq_zero, Finsupp.single_zero]
+
+theorem single_apply {a a' : G} {b : k} [Decidable (a = a')] :
+    single a b a' = if a = a' then b else 0 :=
+  Finsupp.single_apply
+
+theorem single_eq_zero {a : G} {b : k} : single a b = 0 ↔ b = 0 := by
+  simp only [ofFinsupp_eq_zero, Finsupp.single_eq_zero]
+
+theorem toFinsupp_single' : single = (fun (a : G) (b : k) ↦ ofFinsupp (Finsupp.single a b)) := rfl
+
+/-- Group isomorphism between `SkewMonoidAlgebra k G` and `G →₀ k`. This is an
+implementation detail, but it can be useful to transfer results from `Finsupp`
+to `SkewMonoidAlgebra`. -/
+@[simps apply symm_apply]
+def toFinsuppIso : SkewMonoidAlgebra k G ≃+ (G →₀ k) where
+  toFun        := toFinsupp
+  invFun       := ofFinsupp
+  left_inv     := fun ⟨_p⟩ => rfl
+  right_inv _p := rfl
+  map_add'     := toFinsupp_add
+
+theorem smul_single {S} [SMulZeroClass S k] (s : S) (a : G) (b : k) :
+    s • single a b = single a (s • b) :=
+  toFinsupp_injective <| by simp;
+
+theorem single_injective (a : G) : Function.Injective (single a : k → SkewMonoidAlgebra k G) :=
+  (toFinsuppIso).symm.injective.comp (Finsupp.single_injective a)
+
+end Single
+
+section One
+
+variable [One G] [One k]
+
+/-- The unit of the multiplication is `single 1 1`, i.e. the function that is `1` at `1` and
+  zero elsewhere. -/
+instance one : One (SkewMonoidAlgebra k G) :=
+   ⟨⟨Finsupp.single 1 1⟩⟩
+
+@[simp]
+theorem ofFinsupp_one : (⟨Finsupp.single 1 1⟩ : SkewMonoidAlgebra k G) = 1 :=
+  rfl
+
+@[simp]
+theorem toFinsupp_one : (1 : SkewMonoidAlgebra k G).toFinsupp = Finsupp.single 1 1 :=
+  rfl
+
+@[simp]
+theorem toFinsupp_eq_one {a : SkewMonoidAlgebra k G} :
+    a.toFinsupp = Finsupp.single 1 1 ↔ a = 1 := by
+  rw [← toFinsupp_one, toFinsupp_inj]
+
+@[simp]
+theorem ofFinsupp_eq_one {a} :
+    (⟨a⟩ : SkewMonoidAlgebra k G) = 1 ↔ a = Finsupp.single 1 1 := by
+  rw [← ofFinsupp_one, ofFinsupp_inj]
+
+-- This is not a `simp` lemma as `single_zero_left` is more general.
+theorem single_zero_one  : single 1 (1 : k) = 1 :=
+  rfl
+
+theorem one_def : (1 : SkewMonoidAlgebra k G) = single 1 1 :=
+  rfl
+
+end One
+
+section sum
+
+instance [DecidableEq G] [DecidableEq k] : DecidableEq (SkewMonoidAlgebra k G) :=
+  @Equiv.decidableEq (SkewMonoidAlgebra k G) _ (toFinsuppIso).toEquiv (Finsupp.instDecidableEq)
+
+/-- `sum f g` is the sum of `g a (f a)` over the support of `f`. -/
+def sum {N : Type*} [AddCommMonoid N] (f : SkewMonoidAlgebra k G) (g : G → k → N) : N :=
+  (f.toFinsupp).sum g
```

Comment:
Is there much point to this definition? I would think you don't usually see sums indexed by an element of a skew monoid algebra in paper maths

Reply:
This definition is similar to `Finsupp.sum` (as you can see in the def) and indexes the sum on the support of `f`. In my opinion this is very useful and we will need this.

That being said, this definition might change depending on whether or not we do a `coeff` function (see [here](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2310541.3A.20SkewMonoidAlgebra/near/452575660)). Because of this, I will leave this unresolved for now.

Reply:
This has changed, the definition is now very close to the one found in `Polynomial`.